rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -1750,10 +1702,9 @@ extern "C"
             {
                 void *data = guard_conditions->guard_conditions[i];
                 GuardCondition *guard_condition = (GuardCondition*)data;
-                if(!guard_condition->getHasTriggered())
-                {
-                    guard_conditions->guard_conditions[i] = 0;
-                }
+                // Guard conditions persist (don't set to null)
+                // We call getHasTriggered to reset the trigger state.
```

Comment:
I can wait until the next meeting. I think if I focus on the rcl/rclcpp refactor for a bit it will help me understand what the rcl_wait/rmw_wait and guard conditions part of the code should look like. Merging C services will also make working on this easier, since I'll only have to rebase on one set of changes instead of 2.